### PR TITLE
Make debugger and storage stateless

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️

This is the alternative to https://github.com/yiisoft/yii-debug/pull/39

### CODE REVIEW

**Comment by samdark** [comment]
[Debug](https://trello.com/c/r3CdJUaP/1-debug)

File: src/Storage/MemoryStorage.php [hunk 1]
```diff
@@ -3,6 +3,7 @@
 namespace Yiisoft\Yii\Debug\Storage;
 
 use Yiisoft\Yii\Debug\Collector\CollectorInterface;
+use Yiisoft\Yii\Debug\DebuggerIdGenerator;
```
Comment (samdark) [inline src/Storage/MemoryStorage.php]: It isn't used. Should it be?

File: tests/Storage/MemoryStorageTest.php [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 namespace Yiisoft\Yii\Debug\Tests\Storage;
 
+use Yiisoft\Yii\Debug\DebuggerIdGenerator;
```
Comment (samdark) [inline tests/Storage/MemoryStorageTest.php]: It isn't used. Should it be?

Comment (yiiliveext) [reply tests/Storage/MemoryStorageTest.php]: Done

File: src/Collector/CollectorTrait.php [hunk 3]
```diff
@@ -13,6 +13,7 @@ public function startup(): void
 
     public function shutdown(): void
     {
+        $this->reset();
```
Comment (xepozz) [inline src/Collector/CollectorTrait.php]: The trait should provide the abstract method `reset`

File: src/DebuggerIdGenerator.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Yiisoft\Yii\Debug;
+
+class DebuggerIdGenerator
```
Comment (xepozz) [inline src/DebuggerIdGenerator.php]: How about rename generator? It's stateful class that provides generated value. Usually generators provides `generate` method and are stateless.

File: src/DebuggerIdGenerator.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        $this->generateId();
+    }
+
+    public function getId(): string
+    {
+        return $this->id;
+    }
+
+    public function reset(): void
```
Comment (xepozz) [inline src/DebuggerIdGenerator.php]: Why is this needed?

File: src/Storage/FileStorage.php [hunk 6]
```diff
@@ -49,13 +44,15 @@ public function getData(): array
...
-        $this->collectors = [];
+        try {
+            $varDumper = VarDumper::create($this->getData());
+            $jsonData = $varDumper->asJson();
+            $this->filesystem->write($this->path . '/' . $this->idGenerator->getId() . '.data.json', $jsonData);
+
+            $jsonObjects = $varDumper->asJsonObjectsMap();
+            $this->filesystem->write($this->path . '/' . $this->idGenerator->getId() . '.obj.json', $jsonObjects);
+        } finally {
+            $this->collectors = [];
```
Comment (xepozz) [inline src/Storage/FileStorage.php]: I think now is no needs to empty the collectors, because you do `reset()` to all collectors

Comment (yiiliveext) [reply src/Storage/FileStorage.php]: No, they will be added on the next debugger start.

File: src/DebuggerIdGenerator.php (hunk 5 above)
Comment (yiiliveext) [reply src/DebuggerIdGenerator.php]: This is for reset id when the debugger restarts in the Roadrunner environment.

**Comment by samdark** [comment]
Thanks!

File: src/Storage/FileStorage.php (hunk 6 above)
Comment (xepozz) [reply src/Storage/FileStorage.php]: Why?
